### DISCUSSION TITLE
Running CPU-bound blocking work in a command

### DISCUSSION THREAD

**Discussion by mattico** [tauri-apps/tauri#10329] [Q&A]
If I understand the docs correctly, this is not good because it will block the main thread:

```rust
#[tauri::command]
fn foo() {
    std::thread::sleep(Duration::from_millis(1000));
}
```

From reading the docs I thought this was the right thing to do:

```rust
#[tauri::command(async)]
fn foo() {
    std::thread::sleep(Duration::from_millis(1000));
}
```

But looking at the implementation that seems to just use `spawn()`, not `spawn_blocking()`. So I suppose the correct thing to do is to use `spawn_blocking()` explicitly:

```rust
#[tauri::command]
async fn foo() {
    await tauri::async_runtime::spawn_blocking(|| std::thread::sleep(Duration::from_millis(1000)));
}
```

Is that the right thing to do? I think the docs could explain this a little better.

**Comment by FabianLars** [comment]
We generally consider a plain async command good enough, even if not 100% correct. If we're talking about long-ish non-async blocking work that doesn't need to return directly from the command, spawn_blocking or an std thread would indeed be more correct (though i'd still try and check if not using those works without issues 🙃)

**Reply by mattico** [reply]
So I guess the difference is that with e.g. an async web server if you block all your worker threads you can't handle any more requests, and all but N of your requests can't progress. But with a Tauri app, the main thread is the important thing to avoid blocking, really, and the worker threads are separate.

**Reply by FabianLars** [reply]
Yup 👍

**Comment by jacobmichels** [comment]
Stumbled on this thread with the same question as OP. Would be cool to see `#[tauri::command(async_blocking)]` to have a function run with `spawn_blocking` rather than `spawn`.

**Reply by FabianLars** [reply]
Honestly i feel like this is how command(async) on a sync fn should behave. If you want `spawn` then you may as well just use `async fn`. Can't think of a usecase for sync fn + spawn

**Reply by pronebird** [reply]
Had to learn it hard way that `command(async)` does not spawn_blocking. Maybe we should consider adding `async_blocking`? The problem with workarounds that `State<'_, ..>` comes with lifetime so it's difficult to move it into `tauri::async_runtime::spawn_blocking`
